src/item_collaborative_filtering.py

```diff
@@ -23,53 +23,98 @@ def __generate_item_pairs(self, df, item):
+        :param df:  dataframe with columns [user_id, item_id]
+        :return: dictionary with key=item_id, value = interaction probability (float)
+        """
+        item_probabilities = (
+                df.groupby(
+                    self.item_column
+                ).agg(
+                    {self.user_column: 'nunique'}
+                )/df[self.user_column].nunique()
+        ).to_dict()[self.user_column]
 
-        common_users_count = len(item1_users.intersection(item2_users))
+        self.item_probabilities = item_probabilities
 
-        return item_pair, common_users_count
+    def __item_interaction_probability(self, item):
+        """
+        Looks up the probability of a user interacting with an item (e.g. watching, purchasing, liking)
+        :param item: item_id
+        :return: float, probability
+        """
+        try:
+            return self.item_probabilities[item]
+        except AttributeError:
+            "Extract item probabilities first, using __item_probabilities()"
 
-    def __item_interaction_probability(self, df, item):
+    def __user_interactions(self, df):
         """
-        Computes the probability of a user interacting with an item (e.g. watching, purchasing, liking)
+        For ALL users computes the number of items that every user interacted with, APART from item in the df row
         :param df: dataframe with columns [user_id, item_id]
-        :param item:
-        :return: float, probability
+        :return: dictionary with key = user_id, value = number of interactions
         """
-        return df.loc[df[self.item_column]==item, self.user_column].nunique()/df[self.user_column].nunique()
+
+        # subtract 1 to count number of interactions with other items DIFFERENT from item
+        interactions_count = df.groupby(
+            self.user_column
+        )[self.item_column].agg('nunique') - 1
+
+        self.user_interactions = interactions_count.to_dict()
 
     def __count_users_interactions(self, df, item):
         """
-        For ALL users of an item, it computes the number of other items that every user interacted with, APART from item
+        For ALL users of an item, it looks up the number of other items that every user interacted with, APART from item
         :param df: dataframe with columns [user_id, item_id]
         :param item:
-        :return: array of integers = number of interactions. length = nr of users that interacted with item.
+        :return: array of integers with number of interactions per user. length = nr of users that interacted with item.
         """
 
-        filtered_users = df.loc[df[self.item_column] == item, self.user_column]
-
-        # subtract 1 to count number of interactions with other items DIFFERENT from item
-        interactions_count = df[df[self.user_column].isin(filtered_users)].groupby(
-            self.user_column,
-            group_keys=False
```

Comment:
What happened to this one? Why did it fall out?

Reply:
It's still there, I just moved it inside the `try: except` statement